Login Lockout - Module Implementation

arulkumar9690 opened the pull request:
SUMMARY
Login Lockout - Module implementation

ISSUE TYPE
New Module Pull Request

COMPONENT NAME
sonic_login_lockout

ADDITIONAL INFORMATION

Regression test HTML report:
[Login_Lockout_Regression_logs.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/15218836/Login_Lockout_Regression_logs.pdf)

ArunSaravananBalachandran commented on plugins/modules/sonic_login_lockout.py:
```diff
@@ -0,0 +1,175 @@
+DOCUMENTATION = """
+---
+module: sonic_login_lockout
+version_added: '4.2.0'
```
Please change the collection version
```suggestion
version_added: '2.4.0'
```

kerry-meyer replied:
Please use 2.5.0 (here and other applicable places) as the version number for the enterprise_sonic Ansible resource module collection.

arulkumar9690 replied:
Accepted. Done with the changes

ArunSaravananBalachandran commented on plugins/modules/sonic_login_lockout.py:
```diff
@@ -0,0 +1,175 @@
+---
+module: sonic_login_lockout
+version_added: '4.2.0'
+short_description: Manage Global LLDP configurations on SONiC
```
Please update the module name in the description.

kerry-meyer replied:
Also, use 2.5.0 instead of 4.2.0 as the "version_added" value.

arulkumar9690 replied:
Accepted. Done with the changes

ArunSaravananBalachandran commented on plugins/modules/sonic_login_lockout.py:
```diff
@@ -0,0 +1,175 @@
+version_added: '4.2.0'
+short_description: Manage Global LLDP configurations on SONiC
+description:
+  - This module provides configuration management of login lockout parameters.
+    Login Lockout feature is to lock out the user account for user-lockout-period
+    after the max-retry failed attempts. Console exempt option can be enabled 
+    to skip the login lockout validations for console users.
```
Please specify multiple lines as individual lines.

Example:
```yaml
  - This module provides configuration management of login lockout parameters.
  - Login Lockout feature is to lock out the user account for user-lockout-period after the max-retry failed attempts.
```

arulkumar9690 replied:
Accepted. Done with the changes.

ArunSaravananBalachandran commented on plugins/modules/sonic_login_lockout.py:
```diff
@@ -0,0 +1,175 @@
+    description: The set of login lockout attribute configurations
+    type: dict
+    suboptions:
+      console-exempt:
+        description:
+          - Exempt console logins from account lockout.
+        type: bool
+      period:
+        description:
+          - Account lockout period in minutes
+          - The range is from 0 to 43200
+        type: int
+      max-retries:
```
Please use 'underscore' in option names instead of 'hyphen'.

ArunSaravananBalachandran commented on plugins/modules/sonic_login_lockout.py:
```diff
@@ -0,0 +1,175 @@
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community',
+    'license': 'Apache 2.0'
+}
```
Can be removed.

arulkumar9690 replied:
Accepted. Done with the changes

ArunSaravananBalachandran reviewed (changes requested):
Please include the following changes:
- Replaced and Overridden state support
- UT cases

kerry-meyer commented:
Please make the changes requested by Arun and also make the changes needed to fix the flagged sanity errors.

arulkumar9690 commented:
> Please include the following changes:
> 
> * Replaced and Overridden state support
> * UT cases

Accepted. Added Replaced and Overridden support and Added UT cases.

arulkumar9690 commented:
Log Files:
Regression Test Logs : 
[Ansible_Login_Lockout_Regression_logs.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/15164422/Ansible_Login_Lockout_Regression_logs.txt)


UT & Coverage Test Logs:
[Ansible_Login_Lockout_UT_Coverage.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/15164419/Ansible_Login_Lockout_UT_Coverage.log)

ArunSaravananBalachandran commented on tests/unit/modules/network/sonic/fixtures/sonic_login_lockout.yaml:
```diff
@@ -0,0 +1,85 @@
+  expected_config_requests:
+    - path: "data/openconfig-system:system/openconfig-system-ext:login/lockout/config"
+      method: "delete"
+replaced_01:
+  module_args:
+    config:
+      max_retries: 4
+      console_exempt: true
+  state: replaced
```
```suggestion
replaced_01:
  module_args:
    config:
      max_retries: 4
      console_exempt: true
    state: replaced
```

ArunSaravananBalachandran commented on tests/unit/modules/network/sonic/fixtures/sonic_login_lockout.yaml:
```diff
@@ -0,0 +1,85 @@
+          openconfig-system-ext:config:
+            max-retry: 4
+            console-exempt: true
+overridden_01:
+  module_args:
+    config:
+      period: 4
+      console_exempt: true
+      max_retries: 1
+  state: overridden
```
```suggestion
overridden_01:
  module_args:
    config:
      period: 4
      console_exempt: true
      max_retries: 1
    state: overridden
```

arulkumar9690 commented on plugins/module_utils/network/sonic/config/login_lockout/login_lockout.py:
```diff
@@ -0,0 +1,268 @@
+            url = self.login_lockout_config_path['console_exempt']
+            requests.append({'path': url, 'method': DELETE})
+
+        if 'max-retries' in command:
```
Done with the changes

arulkumar9690 commented on plugins/modules/sonic_login_lockout.py:
```diff
@@ -0,0 +1,226 @@
+# !
+# login lockout console-exempt
+# !
+# sonic#
```
Accepted. Done with the changes

arulkumar9690 commented on plugins/module_utils/network/sonic/config/login_lockout/login_lockout.py:
```diff
@@ -0,0 +1,268 @@
+        if (state == 'overridden'):
+            commands = update_states(commands, 'overridden')
+        else:
+            commands = update_states(commands, 'replaced')
```
Accepted. Done with the changes

ArunSaravananBalachandran commented on plugins/module_utils/network/sonic/config/login_lockout/login_lockout.py:
```diff
@@ -0,0 +1,272 @@
+            commands, requests = self._state_replaced_overridden(want, have, diff)
+        return commands, requests
+
+    def _state_replaced_overridden(self, want, have, diff):
+        """ The command generator when state is replaced or overridden
+
+        :rtype: A list
+        :returns: the commands necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        commands = []
+        requests = []
+        delete = {}
+        state = self._module.params['state']
+
+        if (state == 'overridden'):
+            commands = diff
+            requests.extend(self.get_modify_specific_login_lockout_param_requests(commands))
+            commands = update_states(commands, 'overridden')
+        else:
+            for key in have:
+                if key not in diff:
+                    delete[key] = have[key]
+            if delete:
+                commands = delete
+                requests.extend(self.get_delete_specific_login_lockout_param_requests(commands))
+            if diff:
+                commands = diff
+                requests.extend(self.get_modify_specific_login_lockout_param_requests(commands))
+            commands = update_states(commands, 'replaced')
+
+        return commands, requests
```
Since this module has a single instance of simple attributes, the result of replaced and overridden state will be same.
Please update the handling accordingly.

ArunSaravananBalachandran reviewed (changes requested):
Please fix the failing UT and regression test cases for this module.

ArunSaravananBalachandran commented on plugins/module_utils/network/sonic/config/login_lockout/login_lockout.py:
```diff
@@ -0,0 +1,274 @@
+                    continue
+            if key not in diff:
+                delete[key] = have[key]
+        if delete:
+            commands = delete
+            requests.extend(self.get_delete_specific_login_lockout_param_requests(commands))
+        if diff:
+            commands = diff
+            requests.extend(self.get_modify_specific_login_lockout_param_requests(commands))
+
+        if (state == 'overridden'):
+            commands = update_states(commands, 'overridden')
+        else:
+            commands = update_states(commands, 'replaced')
```
Please update the deleted commands when state is replaced/overridden
```suggestion
        if delete:
            commands = update_states(delete, 'deleted')
            requests.extend(self.get_delete_specific_login_lockout_param_requests(delete))
        if diff:
            commands.extend(update_states(diff, state))
            requests.extend(self.get_modify_specific_login_lockout_param_requests(diff))
```

ArunSaravananBalachandran commented on plugins/module_utils/network/sonic/config/login_lockout/login_lockout.py:
```diff
@@ -0,0 +1,274 @@
+        commands = []
+        requests = []
+
+        commands = diff
+
```
Can be removed, since commands is expected to be a list.
```suggestion
```

arulkumar9690 replied:
Accepted. Done with the changes

ArunSaravananBalachandran commented on plugins/module_utils/network/sonic/config/login_lockout/login_lockout.py:
```diff
@@ -0,0 +1,274 @@
+        delete = {}
+        state = self._module.params['state']
+
+        for key in have:
+            if key in want:
+                if want[key] == have[key]:
+                    continue
+            if key not in diff:
+                delete[key] = have[key]
```
Please handle cases when the 'default' value is configured and that option is not specified in want ( no deletion should be performed).

arulkumar9690 commented:
Latest Regression Report : 
[regression-2024-05-03-19-38-06.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/15204566/regression-2024-05-03-19-38-06.pdf)

arulkumar9690 commented:
> Please include the following changes:
> 
> * Replaced and Overridden state support
> * UT cases

Done with the changes

arulkumar9690 commented:
> Please make the changes requested by Arun and also make the changes needed to fix the flagged sanity errors.

Made the changes suggested by Arun and fixed the sanity errors.

ArunSaravananBalachandran reviewed (changes requested):
The code changes look good to me.

Have posted minor comments on the documentation of the module.

arulkumar9690 commented:
> The code changes look good to me.
> 
> Have posted minor comments on the documentation of the module.

Accepted. Done with the changes

kerry-meyer reviewed (approved):
The current proposed change set and posted test results look good.

Thank you for implementing the "Login Lockout" module for the Dell enterprise_sonic Ansible resource module collection.

Approved.
